core/lib/multivm/src/versions/vm_fast/vm.rs

```diff
@@ -353,15 +355,20 @@ impl<S: ReadStorage, Tr: Tracer + Default> Vm<S, Tr> {
                         state_diffs: self.compute_state_diffs(),
                     };
 
-                    // Save the pubdata for the future initial bootloader memory building
-                    self.bootloader_state
-                        .set_pubdata_input(pubdata_input.clone());
-
                     // Apply the pubdata to the current memory
                     let mut memory_to_apply = vec![];
 
-                    apply_pubdata_to_memory(&mut memory_to_apply, pubdata_input);
+                    apply_pubdata_to_memory(
+                        &mut memory_to_apply,
+                        pubdata_builder.expect("`pubdata_builder` is required to finish batch"),
+                        &pubdata_input,
+                        self.system_env.version,
+                    );
                     self.write_to_bootloader_heap(memory_to_apply);
+
+                    // Save the pubdata for the future initial bootloader memory building
+                    self.bootloader_state
+                        .set_pubdata_input(pubdata_input.clone());
```

Comment:
Nit: `clone()` isn't necessary here.

Also, since `pubdata_input` is almost only read from the VM, so it could be relatively straightforward to return `pubdata_input` from this function (e.g., by supplying `pubdata: Option<(&dyn PubdataBuilder, &mut PubdataInput)>` to it), and then use it in `finish_batch()`. I'm not sure a similar transform would work for `vm_latest` though, and we'd probably want to keep `BootloaderState` equivalent for both (ideally, it'd make sense to share it).

Reply:
Removed clone.
Not sure I understood the point for returning `pubdata_input`, can you elaborate on why is it better than storing and reading it from `BootloaderState`?